core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/TransformCorrelatedUnnestToLeftUnnest.java

```diff
@@ -0,0 +1,347 @@
+ *                     - Projection
+ *                          mask_symbol <- ordinality IS NOT NULL
+ *                          - Unnest LEFT with ordinality
+ *                               g <- unnest(c1)
+ *                               u <- unnest(c2)
+ *                               replicate: (a, c1, c2, unique)
+ *                               - AssignUniqueId unique
+ *                                    - Input (a, c1, c2)
+ * </pre>
+ */
+public class TransformCorrelatedUnnestToLeftUnnest
+        implements Rule<CorrelatedJoinNode>
+{
+    private static final Pattern<CorrelatedJoinNode> PATTERN = correlatedJoin()
+            .with(nonEmpty(correlation()))
+            .with(filter().equalTo(TRUE_LITERAL))
+            .matching(node -> node.getType() == CorrelatedJoinNode.Type.INNER || node.getType() == CorrelatedJoinNode.Type.LEFT);
+
+    @Override
+    public Pattern<CorrelatedJoinNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(CorrelatedJoinNode correlatedJoinNode, Captures captures, Context context)
+    {
+        // find global aggregation in subquery
+        List<PlanNode> globalAggregations = PlanNodeSearcher.searchFrom(correlatedJoinNode.getSubquery(), context.getLookup())
+                .where(TransformCorrelatedUnnestToLeftUnnest::isGlobalAggregation)
+                .recurseOnlyWhen(node -> node instanceof ProjectNode || isGlobalAggregation(node))
+                .findAll();
+
+        if (globalAggregations.isEmpty()) {
+            return Result.empty();
+        }
+
+        // if there are multiple global aggregations, the one that is closest to the source is the "reducing" aggregation, because it reduces multiple input rows to single output row
+        AggregationNode reducingAggregation = (AggregationNode) globalAggregations.get(globalAggregations.size() - 1);
+
+        // find unnest in subquery
+        Optional<PlanNode> subqueryUnnest = PlanNodeSearcher.searchFrom(reducingAggregation.getSource(), context.getLookup())
+                .where(node -> isCorrelatedUnnest(node, correlatedJoinNode.getCorrelation(), context.getLookup()))
+                .recurseOnlyWhen(node -> node instanceof ProjectNode || isGroupedAggregation(node))
+                .findFirst();
+
+        if (subqueryUnnest.isEmpty()) {
+            return Result.empty();
+        }
+
+        UnnestNode unnestNode = (UnnestNode) subqueryUnnest.get();
+
+        // assign unique id on input rows to restore semantics of aggregations after rewrite
+        PlanNode input = new AssignUniqueId(
+                context.getIdAllocator().getNextId(),
+                correlatedJoinNode.getInput(),
+                context.getSymbolAllocator().newSymbol("unique", BIGINT));
+
+        // pre-project unnest symbols if they were pre-projected in subquery
+        if (!ImmutableSet.copyOf(correlatedJoinNode.getCorrelation()).containsAll(unnestNode.getMappings().stream()
```

Comment:
Not quite. It is supposed to copy a projection that provides input for UnnestNode. So we cannot do it unconditionally, but need to check if there is a  projection below UnnestNode or not.

However, the condition can be simplified to checking if there is a projection (currently it checks if the projection is _necessary_ for the UnnestNode). It will make the code clearer, and if we plan an unnecessary projection, it should be pruned afterwards.